sandbox/libs/authn/src/main/java/org/opensearch/authn/Permission.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.authn;
+
+/**
+ * This is an abstract class that defines the minimum expectations for a permission object.
+ * A permission implementation of this class can implement further functionality or structuring.
+ * A permission needs to entail the action it allows and the resource its performed on.
+ */
+abstract class Permission {
+
+    // If using a string for construction, a delimiter is required to split the string
+    String PERMISSION_DELIMITER;
+
+    // If using string-object permissions, you use the invalid characters for ensuring formatting
+    String[] INVALID_CHARACTERS;
+
+    // An array of the valid actions which a permission can grant the privilege to perform.
+    String[] QUALIFIED_ACTIONS;
+
+    // An array of the available resources which a permission can grant some action to act upon.
+    String[] QUALIFIED_RESOURCES;
+
+    String permissionString;
+
+    // Every permissionString must be resolvable to its constituent parts: <resource>.<action>
+    // These are then stored separately to avoid costly String manipulation.
+
+    String resource;
+
+    String action;
+
+    abstract void Permission(String permission);
+
+    abstract boolean isValidFormat();
```

Comment:
Lets break this apart, create a PermissionFactory, which has a method `public Permission createPermission(String(s)...)`, this method would enforce the rules.  This prevents 'new' permissions from deviating from our preferred format, then lets create another method `createLegacyPermissions(String)` that doesn't enforce the rules so we can easily identify and fix them later.  This eliminates the need for a `isValidFormat` that would only be used on creation.

This allows both paths to be followed, we still get objects that are a `Permission`, but codifies the old vs new naming scheme.

`Permission` would have the resource/action, in the follow up PR you could consider if 'match' or other methods would be added that are relevant to the check permissions scenarios.

Reply:
Gotcha, I will make these changes.

Reply:
Do you want me to try to add match now or should I wait for another PR? What about test cases etc.? I can combine them or separate things out if you want this to be the abstracts classes/interfaces and have concrete implementations in a follow-up.

Reply:
>Do you want me to try to add match now or should I wait for another PR

I would say wait for the next PR so you can cluster those design elements together in there usage.  Your call

Reply:
Sounds good.